File: ts/problems/20-valid-parentheses.ts
export function isValid(s: string): boolean {
  if (s.length < 2) return false;

  const pairs: Record<string, string> = {
    ")": "(",
    "}": "{",
    "]": "[",
  };

  const stack: Array<string> = [];
  for (const char of s) {
    switch (char) {
      case "(":
      case "{":
      case "[":
        stack.push(char);
        continue;
      case ")":
      case "}":
      case "]":
        if (stack.length > 0 && pairs[char] === stack.at(-1)) {
          stack.pop();
          continue;
        }
      default:
        return false;
    }
  }

  return !stack.length;
}

File: go/problems/20_valid_parentheses.go
package main

func isValid(s string) bool {
	if len(s) < 2 {
		return false
	}

	pairs := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}

	var stack []rune

	for _, char := range s {
		switch char {
		case '(', '{', '[':
			stack = append(stack, char)
		case ')', '}', ']':
			if len(stack) < 1 || pairs[char] != stack[len(stack)-1] {
				return false
			}
			stack = stack[:len(stack)-1]
		default:
			return false
		}
	}

	return len(stack) == 0
}
